x-pack/plugins/stack_connectors/server/connector_types/gemini/gemini.ts

```diff
@@ -0,0 +1,202 @@
+        index: [
+          {
+            names: ['.kibana-event-log-*'],
+            allow_restricted_indices: true,
+            privileges: ['read'],
+          },
+        ],
+      },
+    })) as { has_all_requested: boolean };
+
+    if (!privilege?.has_all_requested) {
+      return { available: false };
+    }
+
+    const response = await initDashboard({
+      logger: this.logger,
+      savedObjectsClient: this.savedObjectsClient,
+      dashboardId,
+      genAIProvider: 'Gemini',
+    });
+
+    return { available: response.success };
+  }
+
+  /** Retrieve access token based on the GCP service account credential json file */
+  private async getAccessToken(): Promise<string | null> {
+    let credentials;
+
+    // Validate the service account credentials JSON file input
+    try {
+      credentials = JSON.parse(this.secrets.credentialsJson);
+    } catch (error) {
+      return `Failed to parse credentials JSON file: Invalid JSON format: ${error.message ?? ''}`;
+    }
+
+    try {
+      const auth = new GoogleAuth({
+        credentials,
+        scopes: 'https://www.googleapis.com/auth/cloud-platform',
+      });
+
+      const token = await auth.getAccessToken();
+      if (token) {
+        GeminiConnector.token = token;
+        // Clear any existing timeout
+        clearTimeout(GeminiConnector.tokenExpiryTimeout);
+
+        // Set a timeout to reset the token after 55 minutes (it expires after 60 minutes)
+        GeminiConnector.tokenExpiryTimeout = setTimeout(() => {
+          GeminiConnector.token = null;
+        }, 55 * 60 * 1000);
+      }
+      return token || null;
+    } catch (error) {
+      return `Failed to get access token ${error.message ?? ''}`;
+    }
+  }
+
+  /**
+   * responsible for making a POST request to the Vertex AI API endpoint and returning the response data
+   * @param body The stringified request body to be sent in the POST request.
+   * @param model Optional model to be used for the API request. If not provided, the default model from the connector will be used.
+   */
+  public async runApi({
+    body,
+    model: reqModel,
+    signal,
+    timeout,
+  }: RunActionParams): Promise<RunActionResponse> {
+    // set model on per request basis
+    const currentModel = reqModel ?? this.model;
+    const path = `/v1/projects/${this.gcpProjectID}/locations/${this.gcpRegion}/publishers/google/models/${currentModel}:generateContent`;
+    const token = await this.getAccessToken();
+
+    const requestArgs = {
+      url: `${this.url}${path}`,
+      method: 'post' as Method,
+      data: body,
+      headers: {
+        Authorization: `Bearer ${token}`,
+        'Content-Type': 'application/json',
+      },
+      signal,
+      timeout: timeout ?? DEFAULT_TIMEOUT_MS,
+      responseSchema: RunApiResponseSchema,
+    } as SubActionRequestParams<RunApiResponse>;
+
+    try {
+      if (!GeminiConnector.token) {
+        GeminiConnector.token = (await this.getAccessToken()) as string;
```

Comment:
Please change the storage of the access token from a static property on a class to Elasticsearch storage. This approach is problematic because it means the access token is reused for every request that uses a Gemini connector, regardless of the specific secret associated with that connector. Additionally, the token doesn't persist over restarts and doesn't handle multi-tenancy. Storing it on the class instance is also not viable since the instance is recreated for every request to Vertex, which could lead to potential access token leakage. You should store the access token in Elasticsearch. Otherwise, you’ll need to request a new access token for each request to the Vertex API, which could result in being rate limited.

Reply:
Sure. We support OAuth for ServiceNow. We store the tokens in a dedicated [saved object](https://www.elastic.co/guide/en/kibana/current/saved-objects-api.html) and have tools to handle them. Example: [x-pack/plugins/actions/server/lib/get_oauth_jwt_access_token.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/actions/server/lib/get_oauth_jwt_access_token.ts)

If you follow the path for ServiceNow you can see how we do it. I would advice following the same pattern.

Reply:
sure, thanks for this @stephmilovic

Reply:
Hey @stephmilovic , we have implemented the changes to store access token. Although we are getting Error: Unknown sort field updatedAt while fetching the access token - in the connector_token_client.ts file. On commenting the sort field option in the get function OR changing the sort field to updated_at -   the flow works as expected .